REGRESSION(284780@main) [GTK][WPE][Debug] Build failure with GCC: no match for ‘operator==’ (operand types are ‘unsigned char:2’ and ‘WebCore::RepaintStatus’)

obyknovenius opened the pull request:
#### cbdd3c329804565a084971c4d6f8f9662ab9251f
<pre>
REGRESSION(284780@main) [GTK][WPE][Debug] Build failure with GCC: no match for ‘operator==’ (operand types are ‘unsigned char:2’ and ‘WebCore::RepaintStatus’)
<a href="https://bugs.webkit.org/show_bug.cgi?id=281056">https://bugs.webkit.org/show_bug.cgi?id=281056</a>

Reviewed by Matt Woodrow and Michael Catanzaro.

This is clearly a compiler bug.
Using `this-&gt;m_repaintStatus` forces `gcc` to re-evaluate the type
and fixes the problem.

* Source/WebCore/rendering/RenderLayer.cpp:
(WebCore::RenderLayer::recursiveUpdateLayerPositions):

Canonical link: <a href="https://commits.webkit.org/284922@main">https://commits.webkit.org/284922@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/720e90eeff993b7a0ada7199ce2dc67e514ef2b5

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/70903 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/50315 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/23676 "Built successfully") | [  ~~🛠 wpe~~](https://ews-build.webkit.org/#/builders/5/builds/75007 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🛠 win~~](https://ews-build.webkit.org/#/builders/59/builds/22108 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/73019 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/58114 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/21926 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/5/builds/75007 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 win-tests~~](https://ews-build.webkit.org/#/builders/59/builds/22108 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/73969 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/45731 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/61127 "Passed tests") | [  ~~🧪 api-wpe~~](https://ews-build.webkit.org/#/builders/5/builds/75007 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/42385 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/18561 "Passed tests") | [  ~~🛠 wpe-cairo~~](https://ews-build.webkit.org/#/builders/65/builds/20449 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/64322 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/18922 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/76722 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/87/builds/15135 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/62/builds/18130 "Passed tests") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/63841 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/86/builds/15179 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/61186 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/63793 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/15701 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/88/builds/11880 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/119/builds/5524 "Passed tests") | | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/46116 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/891 "Built successfully") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/47188 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/48471 "Built successfully") | | | 
| | [  ~~🛠 watch-sim~~](https://ews-build.webkit.org/#/builders/45/builds/46930 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
<!--EWS-Status-Bubble-End-->

darinadler commented on Source/WebCore/rendering/RenderBlockFlow.cpp:
```diff
@@ -3799,7 +3799,7 @@ static void setFullRepaintOnParentInlineBoxLayerIfNeeded(const RenderText& rende
     if (!parent->isInline() || !parent->hasLayer())
         return;
-    downcast<RenderLayerModelObject>(*parent).checkedLayer()->setRepaintStatus(RepaintStatus::NeedsFullRepaint);
+    downcast<RenderLayerModelObject>(*parent).checkedLayer()->setRepaintStatus(NeedsFullRepaint);
```
Can we do the opposite fix, leaving this as an `enum class` and adding some type casts?

mcatanzaro reviewed (approved):
Surely this is better than the original code anyway. Templates are great when they're needed, but using them when a parameter would suffice is excessive.

obyknovenius committed:
REGRESSION(284780@main) [GTK][WPE][Debug] Build failure with GCC: no match for ‘operator==’ (operand types are ‘unsigned char:2’ and ‘WebCore::RepaintStatus’)
https://bugs.webkit.org/show_bug.cgi?id=281056

Reviewed by Matt Woodrow and Michael Catanzaro.

This is clearly a compiler bug.
Using `this->m_repaintStatus` forces `gcc` to re-evaluate the type
and fixes the problem.

* Source/WebCore/rendering/RenderLayer.cpp:
(WebCore::RenderLayer::recursiveUpdateLayerPositions):

Canonical link: https://commits.webkit.org/284922@main

mcatanzaro commented on Source/WebCore/rendering/RenderLayer.cpp:
```diff
@@ -1270,7 +1270,7 @@ void RenderLayer::recursiveUpdateLayerPositions(RenderElement::LayoutIdentifier
     else \
         LAYER_POSITIONS_ASSERT(dest == source);
-    UPDATE_OR_VERIFY_STATE_BIT(m_repaintStatus, RepaintStatus::NeedsNormalRepaint);
+    UPDATE_OR_VERIFY_STATE_BIT(this->m_repaintStatus, RepaintStatus::NeedsNormalRepaint);
```
Wow :(

mcatanzaro replied:
So is this happening with GCC 14? If so, I will try to reproduce and report a bug.

csaavedra replied:
Yes.

darinadler replied:
Typically when we have code to work around a compiler bug, we add a comment so that someone well meaning doesn’t just change the code back. Related: do we have this version of GCC in EWS? If not, how are we going to prevent someone from accidentally changing it back?

csaavedra replied:
The affected code seems to run only inside `ASSERT()` calls. We don't currently have a EWS bot with asserts enabled (no Debug assert EWS bots, only in post-commit), but work is underway to enable asserts in the WPE EWS queues, to  prevent problems like this (and assert crashes in Linux ports!) from sneaking in.

mcatanzaro replied:
Tested with GTK port build of 285131@main with the change from this pull request reverted to reintroduce the bug, using gcc-14.2.1-3.fc41. Unfortunately I'm not able to reproduce the problem.

[Instructions here](https://gcc.gnu.org/bugs/) if somebody else wants to try reporting a bug. You just need to collect the preprocessed source file using -save-temps so the bug report only depends on a single file rather than the entire WebKit build system.
